pkg/controllers/cluster/kubeconfig_controller.go

```diff
@@ -0,0 +1,257 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Greenhouse contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package cluster
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/clientcmd"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+
+	greenhouseapis "github.com/cloudoperators/greenhouse/pkg/apis"
+	"github.com/cloudoperators/greenhouse/pkg/apis/greenhouse/v1alpha1"
+	"github.com/cloudoperators/greenhouse/pkg/clientutil"
+)
+
+type KubeconfigReconciler struct {
+	client.Client
+}
+
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=clusters,verbs=get;list;watch
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=organizations,verbs=get;list;watch
+//+kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch
+// +kubebuilder:rbac:groups=greenhouse.sap,resources=cluster-kubeconfigs,verbs=get;list;watch;create;update;patch
+
+// SetupWithManager sets up the controller with the Manager
+func (r *KubeconfigReconciler) SetupWithManager(name string, mgr ctrl.Manager) error {
+	r.Client = mgr.GetClient()
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		Watches(&v1alpha1.Cluster{}, handler.EnqueueRequestsFromMapFunc(sameNameResource)).
+		Watches(&corev1.Secret{}, handler.EnqueueRequestsFromMapFunc(clusterSecretToCluster)).
+		Watches(&v1alpha1.Organization{}, handler.EnqueueRequestsFromMapFunc(r.organizationToClusters)).
+		Watches(&corev1.Secret{}, handler.EnqueueRequestsFromMapFunc(r.organizationSecretToClusters)).
+		For(&v1alpha1.ClusterKubeconfig{}).
+		Complete(r)
+}
+
+func (r *KubeconfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx).WithValues("cluster", req.Name, "namespace", req.Namespace)
+
+	var cluster v1alpha1.Cluster
```

Comment:
On first sight, we have no proper initialization of Status / Conditions in the Reconcile method.
Also it seems we do not report failure, as we do not update/patch status no matter if the Reconciliation failed...
Have a look at [adr-4](https://github.com/cloudoperators/documentation/blob/main/architecture-decision-records/004-greenhouse-resource-status-reporting.md) about how this was initially conceived...

Reply:
You can look at the [helm_controller](https://github.com/cloudoperators/greenhouse/blob/main/pkg/controllers/plugin/helm_controller.go#L115-L121) or the [team_rbac_controller](https://github.com/cloudoperators/greenhouse/blob/main/pkg/controllers/teamrbac/teamrolebinding_controller.go#L101) how they do it for reference...

Reply:
fixed by 16b7fb7af545499f90a2f168752ee2ceb335a60a (conditions initialisation, update with a deferred function)